crates/wasi-common/yanix/src/file.rs

```diff
@@ -230,20 +232,24 @@ pub unsafe fn fstatat<P: AsRef<OsStr>>(dirfd: RawFd, path: P, flags: AtFlag) ->
 pub unsafe fn fstat(fd: RawFd) -> Result<libc::stat> {
     use std::mem::MaybeUninit;
     let mut filestat = MaybeUninit::<libc::stat>::uninit();
-    Error::from_result(libc::fstat(fd, filestat.as_mut_ptr()))?;
+    from_result(libc::fstat(fd, filestat.as_mut_ptr()))?;
     Ok(filestat.assume_init())
 }
 
 /// `fionread()` function, equivalent to `ioctl(fd, FIONREAD, *bytes)`.
 pub unsafe fn fionread(fd: RawFd) -> Result<u32> {
     let mut nread: libc::c_int = 0;
-    Error::from_result(libc::ioctl(fd, libc::FIONREAD, &mut nread as *mut _))?;
-    Ok(nread.try_into()?)
+    from_result(libc::ioctl(fd, libc::FIONREAD, &mut nread as *mut _))?;
+    nread
+        .try_into()
+        .map_err(|_| Error::from_raw_os_error(libc::EOVERFLOW))
```

Comment:
Right. I won't widen the return type to `u64` at this stage, but I'll leave a note (essentially a copy-paste of your excellent explanation) for future consideration should we want to re-visit this. Does it makes sense?

Reply:
Yes. And actually, there's a subtlety I didn't mention. If you call `FIONREAD` on a file, it can overflow the `int`, and at least Linux doesn't check the overflow, but we have code to handle that case in the code that calls this function, and that's what lets us assume that it won't overflow. Fortunately, an `unwrap()` means that if any OS ever does give us a negative number that we don't expect, it won't pass by silently.